### PR TITLE
copy matcher proto and dependencies from Envoy

### PR DESCRIPTION
**Pull request by snowp**
Moves the Matcher proto into the udpa repo along with a few dependencies: TypedExtensionConfig, StringMatcher and RegexMatcher.

### CODE REVIEW

**Comment by snowp** [comment]
@htuch @markdroth

File: udpa/extension/v1/extension.proto
Comment (markdroth) [inline udpa/extension/v1/extension.proto:3]: Should `TypedExtensionConfig` be in the `xds.extension` tree?  Or should it be in `xds.core`?  This isn't itself an extension; it's really a type used to point at an extension, which is a capability that is basically built into core.

**Review by markdroth — CHANGES_REQUESTED** [review]
Thanks for doing this!

These files should all be in the xds tree, not the udpa tree. We're moving away from udpa.

**Comment by htuch** [comment]
Nice, thanks! Yep, agree that this should be in the `xds/` tree and `TypedExtensionConfig` should be core.

**Comment by snowp** [comment]
@htuch @markdroth ptal, i made the moves

Also FYI I dropped reserved fields, deprecated fields, udpa annotations, etc. when copying these over

**Review by htuch — DISMISSED** [review]
LGTM, will defer to Mark before merge. I think we need to port `protoformat` over to this repo at some point @phlax.

**Review by markdroth — COMMENTED** [review]
Just one potential issue about the java package names.

File: xds/core/v3/extension.proto
Comment (snowp) [reply xds/core/v3/extension.proto]: It would break the generated Java code, so we're extending API compatibility to generated code I don't think we can change it?

Fixed the version string

File: xds/core/v3/extension.proto
Comment (htuch) [reply xds/core/v3/extension.proto:19]: Ack

File: xds/core/v3/extension.proto
Comment (htuch) [reply xds/core/v3/extension.proto]: But this is lift-and-shift across repos, so we don't have strict compatibility requirements, for API or code?

Comment (snowp) [reply xds/core/v3/extension.proto]: Sure we could change the package of this new file to be com.github.xds.core.v1, but won't that make this inconsistent with the other packages in here? I was thinking about the idea of changing the java_package of all the protos in xds/ to not include udpa, which affects more than just this file

Comment (markdroth) [reply xds/core/v3/extension.proto]: I think we should do the following:
1. Decide what we want the java package path to be for all of the xds protos.
2. Use the new path for these new files.
3. Decide on a plan to fix the existing files.  If we can just change them without breaking anything, we should do that; otherwise, we need to come up with an alternative deprecate-and-migrate plan.

@ejona86 may have some thoughts on how this affects things for Java.

**Comment by snowp** [comment]
Sorry about all the force pushes, I need to get my aliases right so I sign commits in repos without the pre-commit hooks :)

File: xds/core/v3/extension.proto
Comment (htuch) [reply xds/core/v3/extension.proto]: I think let's go with `com.github.xds.core.v3` and make sure everything in at least the `xds/` tree is consistent.

**Review by htuch — APPROVED** [review]
LGTM, thanks, will defer to Mark for final approval.

File: xds/core/v3/extension.proto
Comment (markdroth) [reply xds/core/v3/extension.proto]: Can we go ahead and change the existing files in the xds tree?  Or will that break existing usage?

**Review by markdroth — APPROVED** [review]
Looks great!  We can follow up on fixing the java package path on the existing files separately.

**Comment by markdroth** [comment]
Harvey, I don't have merge access on this repo, so you'll need to do that.

File: xds/core/v3/extension.proto
Comment (ejona86) [reply xds/core/v3/extension.proto]: What is the migration process expected for this proto? It has a different package name and file path, so it seems like a brand new proto from protobuf's perspective. In that case, it is a new proto with no prior usages and the Java path can be anything but `io.envoyproxy.envoy.config.core.v3` (to avoid colliding with the other copy).

Comment (ejona86) [reply xds/core/v3/extension.proto]: Mark spoke to me and made clear the issue is the existing files.

The existing files are [already included in grpc-java](
https://github.com/grpc/grpc-java/tree/master/xds/third_party/udpa/src/main/proto/xds/core/v3), however, [there aren't any references to the messages](https://github.com/grpc/grpc-java/search?q=com.github.udpa.xds.core.v3) outside of generated code. That means that as long as the consuming protos are generated in the same build process as these protos we'd be fine. That is true for Google-internal and, since grpc-java is not using packages from maven-central but instead building and shading the generated code in grpc-xds. it would be true for the github builds of grpc-java as well.

So it seems fine to change these package names from grpc-java's perspective.

Comment (markdroth) [reply xds/core/v3/extension.proto]: Okay, super, I'll send a separate PR to update the existing files.

Should we use `com.github.cncf.xds` instead of just `com.github.xds`, to avoid potential conflicts with some other github `<something>/xds` repo?

Comment (markdroth) [reply xds/core/v3/extension.proto]: Actually, nevermind that last question -- we're already using `xds` as the top-level proto package, so `com.github.xds` is probably fine.

Comment (markdroth) [reply xds/core/v3/extension.proto]: Sent #9 with the changes to the existing files.

**Comment by htuch** [comment]
@markdroth done, I've fixed it so you should be able to merge in the future.

**Pull request merged**
